src/util/bson/bson.cpp

```diff
@@ -443,641 +444,196 @@ std::ostream& operator<<(std::ostream& out, const Bson& b)
-                m_instructions.push({State::Skip});
-            } else {
-                // we will ignore the subtype for now
-                m_marks.top().push_back(instruction.key, std::vector<char>());
-            }
-
-            break;
-        case State::BinaryBase64: {
-            // if we have already pushed a binary type
-            if (m_marks.top().size() && m_marks.top().back().first == instruction.key) {
-                m_marks.top().pop_back();
-                m_marks.top().push_back(instruction.key, std::vector<char>(val.begin(), val.end()));
-
-                // pop vestigal binary instruction
-                m_instructions.pop();
-                m_instructions.push({State::Skip});
-                m_instructions.push({State::Skip});
-            } else {
-                // we will ignore the subtype for now
-                m_marks.top().push_back(instruction.key, std::vector<char>(val.begin(), val.end()));
-            }
-
-            break;
+            return Bson(std::move(out));
         }
-        default:
-            m_marks.top().push_back(instruction.key, std::string(val.begin(), val.end()));
-            break;
+        case Json::value_t::discarded:
+            REALM_TERMINATE("should never see discarded");
     }
-    return true;
-}
-
-/*!
- @brief an object key was read
- @param[in] val  object key
- @return whether parsing should proceed
- @note It is safe to move the passed string.
- */
-bool Parser::key(string_t& val) {
-    if (!m_instructions.empty()) {
-        auto top = m_instructions.top();
-
-        if (top.type == State::RegularExpression) {
-            if (val == key_regular_expression_pattern) {
-                m_instructions.push({State::RegularExpressionPattern, top.key});
-            } else if (val == key_regular_expression_options) {
-                m_instructions.push({State::RegularExpressionOptions, top.key});
-            }
-            return true;
-        } else if (top.type == State::Date) {
-            return true;
-        } else if (top.type == State::Binary) {
-            if (val == key_binary_base64) {
-                m_instructions.push({State::BinaryBase64, top.key});
-            } else if (val == key_binary_sub_type) {
-                m_instructions.push({State::BinarySubType, top.key});
+    REALM_TERMINATE("unknown json value type");
+}
+
+// This works around the deleted rvalue constructor in StringData
+inline StringData tosd(const std::string& s) { return s; }
+
+// Keep these sorted by key. This is checked so you can't forget.
+using FancyParser = Bson(*)(const Json& json);
+static constexpr std::pair<std::string_view, FancyParser> bson_fancy_parsers[] = {
+    {"$binary", +[](const Json& json) {
+        util::Optional<std::vector<char>> base64;
+        util::Optional<uint8_t> subType;
+        if (json.size() != 2)
+            throw BsonError("invalid extended json $binary");
+        for (auto&& [k, v] : json.items()) {
```

Comment:
Is this syntax destructuring a pair as an rval ref?

Reply:
`auto&&` is just generally the best default thing to use for range-for loops because it always does the right thing. Depending on what is yielded from the `json.items()` range it will either be a `&`, a `const&`, or a `&&` (it could also be a `const&&` but best to pretend they don't exist and behave like `const&`). There are two complicating factors to keep in mind though:

1) names are always lvalues, even if they are rvalue-refs. So you still need `std::move(x)` or `std::forward<decltype(x)>(x)` if you want it to be moved-from/forwarded.
2) the cvref qualifications (const, volatile, reference) on a structured binding generally apply to the hidden variable used to capture the whole object, not to the separate bindings. (The actual mechanics of structured bindings are kinda weird but they generally just behave how you expect, as long as you remember that they are just creating new bindings, not actually breaking up the object, since that would cause lifetime and object model issues)

TL;DR: That is the (mostly) full story. Luckily, you basically never need to think about any of that. Just always use `auto&&` in your range-for loops and write whatever code you would write if it was an alias for the elements in the range.

Reply:
👍 Thanks for the tips.